nixos/modules/services/audio/alsa.nix

```diff
@@ -1,38 +1,439 @@
+    hardware.pulseaudio.enable = false;
+
+    hardware.alsa.config =
+      let
+        conf = [
+          ''
+            pcm.!default fromenv
+
+            # Read the capture and playback device from
+            # the ALSA_AUDIO_IN, ALSA_AUDIO_OUT variables
+            pcm.fromenv {
+              type asym
+              playback.pcm {
+                type plug
+                slave.pcm {
+                  @func getenv
+                  vars [ ALSA_AUDIO_OUT ]
+                  default pcm.sysdefault
+                }
+              }
+              capture.pcm {
+                type plug
+                slave.pcm {
+                  @func getenv
+                  vars [ ALSA_AUDIO_IN ]
+                  default pcm.sysdefault
+                }
+              }
+            }
+          ''
+          (lib.optional cfg.enableRecorder ''
+            pcm.!default "splitter:fromenv,recorder"
+
+            # Send audio to two stereo devices
+            pcm.splitter {
+              @args [ A B ]
+              @args.A.type string
+              @args.B.type string
+              type asym
+              playback.pcm {
+                type plug
+                route_policy "duplicate"
+                slave.pcm {
+                  type multi
+                  slaves.a.pcm $A
+                  slaves.b.pcm $B
+                  slaves.a.channels 2
+                  slaves.b.channels 2
+                  bindings [
+                   { slave a channel 0 }
+                   { slave a channel 1 }
+                   { slave b channel 0 }
+                   { slave b channel 1 }
+                  ]
+                }
+              }
+              capture.pcm $A
+            }
 
+            # Device which records and plays back audio
+            pcm.recorder {
+              type asym
+              capture.pcm {
+                type dsnoop
+                ipc_key 9165218
+                ipc_perm 0666
+                slave.pcm "hw:loopback,1,0"
+                slave.period_size 1024
+                slave.buffer_size 8192
+              }
+              playback.pcm {
+                type dmix
+                ipc_key 6181923
+                ipc_perm 0666
+                slave.pcm "hw:loopback,0,0"
+                slave.period_size 1024
+                slave.buffer_size 8192
+              }
+            }
+          '')
+          (lib.mapAttrsToList mkControl cfg.controls)
+          (lib.mapAttrsToList (n: v: "pcm.${n} ${quote v}") cfg.deviceAliases)
+        ];
+      in
+      lib.mkBefore (lib.concatStringsSep "\n" (lib.flatten conf));
+
+    hardware.alsa.cardAliases = lib.mkIf cfg.enableRecorder {
+      loopback.driver = "snd_aloop";
+      loopback.id = 2;
+    };
+
+    # Set default PCM devices
+    environment.sessionVariables = defaultDeviceVars;
+    systemd.globalEnvironment = defaultDeviceVars;
+
+    environment.etc."asound.conf".text = cfg.config;
+
+    boot.kernelModules =
+      [ ]
```

Comment:
```suggestion
```

Reply:
It's a pretty standard trick to keep the entries aligned and the diff small.